### PR TITLE
Fix function.apply type definition

### LINKED ISSUE #22600: Type Definition for function.apply is wrong
**TypeScript Version:**  2.7.2

**Search Terms:** 
apply lib

**lib.es5.d.ts:**
**lib.es6.d.ts:**
```ts
apply(this: Function, thisArg: any, argArray?: any): any;
```
`argArray` actually accepts an array of arguments but since it's defined as any, you can pass in non array types which lead to a TypeError at runtime.

**Code**

```ts
const foo = (a) => alert(a);

try {
    foo.apply(foo, "not array");    //TypeError
}
catch(e) {
    alert(e);
}
foo.apply(foo, ["in array"]);    //ok
```

**Expected behavior:**
Compile Error since argument is not an array.
**Actual behavior:**
No compiletime error
Triggers `TypeError` at runtime
**Playground Link:**  
http://www.typescriptlang.org/play/index.html#src=const%20foo%20%3D%20(a)%20%3D%3E%20alert(a)%3B%0D%0A%0D%0Atry%20%7B%0D%0A%20%20%20%20foo.apply(foo%2C%20%22not%20array%22)%3B%20%20%20%20%2F%2FTypeError%0D%0A%7D%0D%0Acatch(e)%20%7B%0D%0A%20%20%20%20alert(e)%3B%0D%0A%7D%0D%0Afoo.apply(foo%2C%20%5B%22in%20array%22%5D)%3B%20%20%20%20%2F%2Fok

### PR DESCRIPTION
**Pull request by adriengibrat**
Fixes #22600

### CODE REVIEW

File: lib/lib.d.ts [hunk 1]
```diff
@@ -265,7 +265,7 @@ interface Function {
       * @param thisArg The object to be used as the this object.
       * @param argArray A set of arguments to be passed to the function.
       */
-    apply(this: Function, thisArg: any, argArray?: any): any;
+    apply(this: Function, thisArg: any, argArray?: any[]): any;
```
Comment (DanielRosenwasser) [inline lib/lib.d.ts]: Is this compatible with `IArguments`? What about `ReadonlyArray`?

**Comment by DanielRosenwasser** [comment]
I think you really want a readonly-version of `ArrayLike`. In other words, instead of `any[]`, you want `Readonly<ArrayLike<any>>`.

**Comment by adriengibrat** [comment]
Thanks for your review, using `Readonly<ArrayLike<any>>` is indeed way better, done in b489fdcd36

**Comment by RyanCavanaugh** [comment]
@DanielRosenwasser 👍 ? CircleCI failures are unrelated

**Comment by adriengibrat** [comment]
I can rebase to upstream/master and push to run circle ci again?

**Comment by mhegazy** [comment]
thanks @adriengibrat!

**Pull request merged**